plonk-core/src/proof_system/pi.rs

```diff
@@ -0,0 +1,193 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) ZK-Garage. All rights reserved.
+
+//! Public Inputs of the circuit. This values are available for the
+//! [`super::Prover`] and [`super::Verifier`].
+//!
+//! This module contains the implementation of the [`PI`] struct and all the
+//! basic manipulations such as inserting new values and getting the public
+//! inputs in evaluation or coefficient form.
+
+use alloc::collections::BTreeMap;
+use ark_ff::{FftField, ToConstraintField};
+use ark_poly::{
+    polynomial::UVPolynomial, univariate::DensePolynomial, EvaluationDomain,
+    GeneralEvaluationDomain,
+};
+use ark_serialize::{
+    CanonicalDeserialize, CanonicalSerialize, Read, SerializationError, Write,
+};
+
+use crate::prelude::Error;
+
+///  Public Inputs
+#[derive(
+    CanonicalDeserialize, CanonicalSerialize, Clone, PartialEq, Debug, Hash,
+)]
+pub struct PI<F>
+where
+    F: FftField,
+{
+    // padded size of the circuit
+    n: usize,
+    // non-zero values of the public input
+    values: BTreeMap<usize, F>,
+}
+
+impl<F> PI<F>
+where
+    F: FftField,
+{
+    /// Creates a new struct for [`PI`].
+    pub fn new(n: usize) -> Self {
+        assert!(n.is_power_of_two());
+        Self {
+            n,
+            values: BTreeMap::new(),
+        }
+    }
+
+    /// Updates the size of the circuit.
+    pub fn update_size(&mut self, n: usize) {
+        assert!(n.is_power_of_two());
+        self.n = n;
+    }
+
+    /// Inserts a new public input value at a given position.
+    ///
+    /// If the value provided is zero no insertion will be made as zeros are the
+    /// implicit value of empty positions.
+    /// The function will panic if an insertion is atempted in an already
+    /// occupied position.
+    pub fn insert(&mut self, pos: usize, val: F) {
+        if self.values.contains_key(&pos) {
+            panic!("Insertion in public inputs conflicts with previous value at position {}", pos);
+        }
+        if val != F::zero() {
+            self.values.insert(pos, val);
+        }
+    }
```

Comment:
This does not prevent someone from doing the following:

```rust
self.insert(4, 0);
self.insert(4, 1);
```

or in the other order. Do we even want to allow this?